File: src/main/java/dev/turtlebongo/ThatGuyJustin/ServerTools/ServerTools.java
package dev.turtlebongo.ThatGuyJustin.ServerTools;

import dev.turtlebongo.ThatGuyJustin.ServerTools.discord.DiscordHandler;
import dev.turtlebongo.ThatGuyJustin.ServerTools.util.Logger;
import dev.turtlebongo.ThatGuyJustin.ServerTools.util.StringUtils;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.exceptions.InsufficientPermissionException;
import net.minecraft.network.chat.*;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.phys.Vec3;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.ServerChatEvent;
import net.minecraftforge.event.server.ServerStartedEvent;
import net.minecraftforge.event.server.ServerStoppingEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.ModLoadingContext;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.config.ModConfig;
import net.minecraftforge.event.server.ServerStartingEvent;
import net.minecraftforge.server.ServerLifecycleHooks;

import java.io.File;
import java.util.*;

// The value here should match an entry in the META-INF/mods.toml file
@Mod("servertools")
public class ServerTools
{

    private DiscordHandler discordHandler;
    private Thread timer;
    private Date startup = new Date();
    private List<UUID> playerLoginCache = new ArrayList<>();
    private HashMap<UUID, Vec3> locationCache = new HashMap<>();
    private HashMap<String, List<String>> mob_filter = new HashMap<>();


    public ServerTools()
    {
        // Register ourselves for server and other game events we are interested in
        MinecraftForge.EVENT_BUS.register(this);

        ModLoadingContext.get().registerConfig(ModConfig.Type.COMMON, Config.GENERAL_SPEC, "servertools.toml");
    }

    // You can use SubscribeEvent and let the Event Bus discover methods to call
    @SubscribeEvent
    public void onServerStarting(ServerStartingEvent event)
    {
        // Do something when the server starts
        if(Config.discordEnabled.get()) {
            Logger.info("Starting Discord Handler...", true);
            this.discordHandler = new DiscordHandler(this.startup);
        }

    }

    @SubscribeEvent
    public void onServerStarted(ServerStartedEvent event){
        File server_dir = ServerLifecycleHooks.getCurrentServer().getServerDirectory();
        this.timer = new Thread(() -> {
            try {

                MinecraftServer server = ServerLifecycleHooks.getCurrentServer();
                Logger.info("Timer started!", true);
//                sleepUntilBroadcast(10000, buildAnnouncement("in 1 Hour     ", "#c55d5d"));
//                sleepUntilBroadcast(10000, buildAnnouncement("in 30 Minutes ", "#c65958"));
//                sleepUntilBroadcast(10000, buildAnnouncement("in 10 Minutes ", "#c85553"));
//                sleepUntilBroadcast(10000, buildAnnouncement("in 5 Minutes  ", "#c9504e"));
//                sleepUntilBroadcast(10000, buildAnnouncement("in 1 Minute   ", "#c94c49"));
//                sleepUntilBroadcast(10000, buildAnnouncement("in 10 Seconds ", "#ca4744"));
                int rootHours = Config.restartHours.get();
                sleepUntilBroadcast((long) (rootHours - 1) * 60 * 60 * 1000, buildAnnouncement("in 1 Hour", "#fcff00"));
                sleepUntilBroadcast(300 * 60 * 1000, buildAnnouncement("in 1 Hour", "#fcff00"));
                sleepUntilBroadcast(30 * 60 * 1000, buildAnnouncement("in 30 Minutes", "#fbed00"));
                sleepUntilBroadcast(20 * 60 * 1000, buildAnnouncement("in 10 Minutes", "#f9dc00"));
                sleepUntilBroadcast(5 * 60 * 1000, buildAnnouncement("in 5 Minutes", "#f5ca00"));
                sleepUntilBroadcast(4 * 60 * 1000, buildAnnouncement("in 1 Minute", "#f0b900"));
                sleepUntilBroadcast(50000, buildAnnouncement("in 10 Seconds", "#eba800"));
                String[] colors = { "#ca0000", "#cb0f0a", "#cb1913", "#cc211b", "#cc2821", "#cc2e28", "#cc332e", "#cc3934", "#cb3e39", "#cb433f"};
                int countdown = 9;
                while(countdown > 0){
                    sleepUntilBroadcast(1000, buildAnnouncement("in " + countdown + " Seconds  ", colors[countdown]));
                    countdown--;
                }
                server.sendSystemMessage(buildAnnouncement("RIGHT NOW ", colors[0]));
                ServerTools.shutdownServer();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });
        timer.setName("Server Shutdown Timer");
        timer.start();
        this.parseValues();
    }

    @SubscribeEvent
    public void onServerShutdown(ServerStoppingEvent event){
        Logger.info("Shutting down all handlers...", true);
//        this.timer.stop();
        if(Config.discordEnabled.get()) {
            if (Config.loggingChannel.get() != null) {
                TextChannel logs = this.discordHandler.getBotClient().getTextChannelById(Config.loggingChannel.get());
                String msg = String.format("[<t:%s:T>] Server is shutting down.",
                        new Date().getTime() / 1000);
                try {
                    logs.sendMessage(msg).queue();
                } catch (InsufficientPermissionException e) {
                    Logger.error("Unable to post in log channel:", true);
                    e.printStackTrace();
                }
            }
            this.discordHandler.shutdown();
        }

    }

//    @SubscribeEvent
//    public void onLogin(PlayerEvent.PlayerLoggedInEvent event){
//        try {
////            event.getPlayer().setInvulnerable(true);
//            this.playerLoginCache.add(event.getPlayer().getUUID());
//            this.locationCache.put(event.getPlayer().getUUID(), event.getPlayer().getPosition(0));
//        }catch (Exception e){
//        }
//    }

//    @SubscribeEvent
//    public void onLogoff(PlayerEvent.PlayerLoggedOutEvent event){
//        if(this.playerLoginCache.contains(event.getPlayer().getUUID())){
//            event.getPlayer().setInvulnerable(false);
//            this.playerLoginCache.remove(event.getPlayer().getUUID());
//        }
//    }
//
//    @SubscribeEvent
//    public void onPlayerInteract(PlayerInteractEvent event){
//        if(this.playerLoginCache.contains(event.getPlayer().getUUID())){
//            event.getPlayer().getPosition(0);
//            event.getPlayer().setInvulnerable(false);
//            this.playerLoginCache.remove(event.getPlayer().getUUID());
//            this.locationCache.remove(event.getPlayer().getUUID());
//        }
//    }
//
//    @SubscribeEvent
//    public void onEntityUpdate(LivingEvent.LivingUpdateEvent event){
//        if(this.locationCache.containsKey(event.getEntity().getUUID())){
//            if(!this.locationCache.containsKey(event.getEntity().getUUID())) return;
//            if(!this.locationCache.get(event.getEntity().getUUID()).equals(event.getEntity().getPosition(0))){
//                this.locationCache.remove(event.getEntity().getUUID());
//                event.getEntity().setInvulnerable(false);
//                this.playerLoginCache.remove(event.getEntity().getUUID());
//            }
//        }
//    }

    @SubscribeEvent
    public void onChatMessage(ServerChatEvent event){
//        MutableComponent finalFormat = ChatFormatting.format(event);
        if(!Config.discordEnabled.get()) return;
        if(!Config.enabledChatBridge.get()) return;
        this.discordHandler.sendWebHookMessage(event.getPlayer(), event.getMessage().getString());
    }

//    @SubscribeEvent
//    public void onEntitySpawn(LivingSpawnEvent event){
//        if(Config.mob_filter_enabled.get()) {
//            if (this.mob_filter.containsKey(event.getEntity().getType().getRegistryName().toString())) {
//                String world_name = event.getEntity().level.dimension().location().toString();
//                if (this.mob_filter.get(event.getEntity().getType().getRegistryName().toString()).contains(world_name)) {
////                event.setCanceled(true);
//                    event.getEntity().remove(Entity.RemovalReason.DISCARDED);
//                }
//            }
//        }
//    }

//    @SubscribeEvent
//    public static void RegisterCommands(RegisterCommandsEvent event) {
//        RegisterSlashCommands.register(event.getDispatcher());
//
//    }

    public static double getTPS() {
        double meanTickTime = ServerLifecycleHooks.getCurrentServer().getAverageTickTime() * 1.0E-6D;
        return Math.min(1000.0 / meanTickTime, 20);
    }

    public static void shutdownServer(){
        MinecraftServer server = ServerLifecycleHooks.getCurrentServer();
        server.getCommands().performPrefixedCommand(server.createCommandSourceStack(), "stop");
    }

    private void sleepUntilBroadcast(long time, Component msg) throws InterruptedException {
        MinecraftServer server = ServerLifecycleHooks.getCurrentServer();
        Thread.sleep(time);
//        server.getPlayerList().broadcastMessage(msg, ChatType.SYSTEM, UUID.randomUUID());
//        server.getPlayerList().broadcastChatMessage(msg, server.createCommandSourceStack(), ChatType.);
        server.sendSystemMessage(msg);
        for (ServerPlayer p : server.getPlayerList().getPlayers()) {
            p.displayClientMessage(msg, false);
        }
    }

    private MutableComponent buildAnnouncement(String timeLeft, String hex){
        Style s = Style.EMPTY.withColor(TextColor.parseColor(hex)).withBold(true);

        Component staticMiddle = Component.literal(StringUtils.color("&7Server Reboot will happen "));
        MutableComponent announce = Component.literal(StringUtils.color("&8(")).append(Component.literal("!").withStyle(s)).append(Component.literal(StringUtils.color("&8) ")));
        MutableComponent time = Component.literal(timeLeft).withStyle(s);
        MutableComponent finalMsg = Component.literal(StringUtils.color("&8(")).append(Component.literal("!").withStyle(s)).append(Component.literal(StringUtils.color("&8) "))).append(staticMiddle).append(time).append(announce);

//        TextComponent staticMiddle = new TextComponent(StringUtils.color("&7Server Reboot will happen "));
//        MutableComponent announce = new TextComponent(StringUtils.color("&8(")).append(new TextComponent("!").withStyle(s)).append(new TextComponent(StringUtils.color("&8) ")));
//        MutableComponent time = new TextComponent(timeLeft).withStyle(s);
//        MutableComponent finalMsg = new TextComponent(StringUtils.color("&8(")).append(new TextComponent("!").withStyle(s)).append(new TextComponent(StringUtils.color("&8) "))).append(staticMiddle).append(time).append(announce);

        return finalMsg;
    }

    private void parseValues(){
        List<? extends String> raw_values = Config.mob_filter.get();
        for(String s: raw_values){
            String mob_name = s.split("/")[0];
            String[] worlds = s.split("/")[1].split(",");
            this.mob_filter.put(mob_name, Arrays.asList(worlds));
        }
    }

}

File: src/main/kotlin/dev/fatyoshi/ThatGuyJustin/servertools/ServerTools.kt
package dev.fatyoshi.thatguyjustin.servertools

import de.maxhenkel.admiral.MinecraftAdmiral
import dev.fatyoshi.thatguyjustin.servertools.discord.DiscordHandler
import dev.fatyoshi.thatguyjustin.servertools.util.Logger
import dev.fatyoshi.thatguyjustin.servertools.util.sendAll
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel
import net.dv8tion.jda.api.exceptions.InsufficientPermissionException
import net.kyori.adventure.platform.modcommon.MinecraftServerAudiences
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.fml.ModLoadingContext
import net.neoforged.fml.common.Mod
import net.neoforged.fml.config.ModConfig
import net.neoforged.neoforge.common.NeoForge
import net.neoforged.neoforge.event.RegisterCommandsEvent
import net.neoforged.neoforge.event.ServerChatEvent
import net.neoforged.neoforge.event.server.ServerStartedEvent
import net.neoforged.neoforge.event.server.ServerStartingEvent
import net.neoforged.neoforge.event.server.ServerStoppedEvent
import net.neoforged.neoforge.event.server.ServerStoppingEvent
import net.neoforged.neoforge.server.ServerLifecycleHooks
import java.util.*
import kotlin.text.get
import kotlin.time.Duration
import kotlin.time.toJavaDuration

@Mod("servertools")
class ServerTools {
    private var adventure: MinecraftServerAudiences? = null
    private val noticesRegex = Regex("""^time=(?<duration>[^,]+),message=(?<message>.*)$""")
    private var discordHandler: DiscordHandler? = null
    private val mobFilter = HashMap<String, List<String>>()

    private var timer: Thread? = null
    var timerActive = false
    var restartDuration: Duration? = null
    var timerStarted: Date? = null
    lateinit var startup: Date

    companion object {
        lateinit var instance: ServerTools
    }

    init {
        NeoForge.EVENT_BUS.register(this);
        ModLoadingContext.get().activeContainer.registerConfig(ModConfig.Type.COMMON, Config.GENERAL_SPEC, "servertools.toml")

        instance = this
    }

    @SubscribeEvent
    fun onServerStarting(event: ServerStartingEvent) {
        this.adventure = MinecraftServerAudiences.of(event.getServer())
    }

    @SubscribeEvent
    fun onCommandReg(event: RegisterCommandsEvent) {
        MinecraftAdmiral.builder(event.dispatcher, event.buildContext).addCommandClasses(
            Commands().javaClass, Uptime().javaClass, Timer().javaClass
        ).build()
    }

    @SubscribeEvent
    fun onServerShutdown(event: ServerStoppedEvent) {
        this.adventure = null
    }


    @SubscribeEvent
    fun onChatMessage(event: ServerChatEvent) {
        if (!Config.discordEnabled!!.get()) return
        if (!Config.enabledChatBridge!!.get()) return
        discordHandler!!.sendWebHookMessage(event.player, event.message.string)
    }

    @SubscribeEvent
    fun onServerStarted(event: ServerStartedEvent) {
        // Set times for initial startup and timer. Mainly used for time deltas.
        this.startup = Date()

        // Start up the discord bot if enabled.
        if (Config.discordEnabled!!.get()) {
            Logger.info("Starting Discord Handler...", true)
            this.discordHandler = DiscordHandler(this.startup!!)
        }

        // Mob Filter.
        this.parseValues()

        // Anddddd start the timer!
        if (Config.enableRestartTimer!!.get() == false) return
        this.restartDuration = Duration.parse(Config.restartTime!!.get())
        startTimer(startup=true)
    }

    @SubscribeEvent
    fun onServerShutdown(event: ServerStoppingEvent?) {
        Logger.info("Shutting down all handlers...", true)
        if(this.timerActive)
            try{
                this.timer!!.interrupt()
            }catch (_: InterruptedException){
                Logger.info("Timer stopped.", true)
            }
        if (Config.discordEnabled!!.get()) {
            if (Config.loggingChannel!!.get() != "") {
                val logs: TextChannel =
                    discordHandler?.botClient?.getTextChannelById(Config.loggingChannel!!.get()) ?: return
                val msg = String.format(
                    "[<t:%s:T>] Server is shutting down.",
                    Date().time / 1000
                )
                try {
                    logs.sendMessage(msg).queue()
                } catch (e: InsufficientPermissionException) {
                    Logger.error("Unable to post in log channel:", true)
                    e.printStackTrace()
                }
            }
            discordHandler!!.shutdown()
        }
    }

    //    @SubscribeEvent
    //    public void onEntitySpawn(LivingSpawnEvent event){
    //        if(Config.mob_filter_enabled.get()) {
    //            if (this.mob_filter.containsKey(event.getEntity().getType().getRegistryName().toString())) {
    //                String world_name = event.getEntity().level.dimension().location().toString();
    //                if (this.mob_filter.get(event.getEntity().getType().getRegistryName().toString()).contains(world_name)) {
    ////                event.setCanceled(true);
    //                    event.getEntity().remove(Entity.RemovalReason.DISCARDED);
    //                }
    //            }
    //        }
    //    }
    fun stopTimer(){
        timerActive = false
        try{
            this.timer!!.interrupt()
        }catch (_: InterruptedException){}
    }

    fun startTimer(time: Duration? = null, startup: Boolean = false) {
        if (startup)
            instance.timerStarted = instance.startup
        else
            instance.timerStarted = Date()
        timer = Thread {
            try {
                val restartIn = time ?: restartDuration

                instance.restartDuration = restartIn

                // Time to make the config more system readable...and sort it...
                val rawMessageTimes = HashMap<Duration, String>()
                Config.restartTimerNotices!!.get().forEach { rawCfgValue ->
                    val match = noticesRegex.find(rawCfgValue)
                    if (match != null) {
                        val rawDuration = match.groups["duration"]?.value
                        val message = match.groups["message"]?.value
                        val duration = Duration.parse(rawDuration!!)
                        if (message != null && duration < restartIn!!) rawMessageTimes[duration] = message
                    }
                }

                // In the end we want it as longest duration to the shortest duration.
                val messageTimes = rawMessageTimes.toSortedMap().reversed().toMap()

                Logger.info("&fStartup Time&7: &a${this.startup}", true)
                if(time != null){
                    Logger.info("&fExpected Restart Time&7: &a${this.timerStarted!!.toInstant().plus(time.toJavaDuration())}", true)
                }else{
                    Logger.info("&fConfigured Restart Time&7: &a${this.timerStarted!!.toInstant().plus(restartIn!!.toJavaDuration())}", true)
                }
                instance.timerActive = true
                Logger.info("Timer started!", true)

                var currentSleep = restartIn

                messageTimes.forEach { (duration, message) ->
                    sleepUntilBroadcast(currentSleep!!.minus(duration).inWholeMilliseconds, message)
                    currentSleep = duration
                }

                instance.timerActive = false
                instance.shutdownServer()
            } catch (e: InterruptedException) {
                if(timerActive)
                    Logger.warning("&cTimer interrupted, is the server restarting sooner?", true)
            }
        }
        timer!!.name = "Server Shutdown Timer"
        timer!!.start()
    }

    fun getTPS(): Double {
        val meanTickTime = ServerLifecycleHooks.getCurrentServer()!!.averageTickTimeNanos * 1.0E-6
        return Math.min(1000.0 / meanTickTime, 20.0)
    }

    fun shutdownServer() {
        val server = ServerLifecycleHooks.getCurrentServer()!!
        server.commands.performPrefixedCommand(server.createCommandSourceStack(), "stop")
    }

    @Throws(InterruptedException::class)
    private fun sleepUntilBroadcast(time: Long, msg: String) {
        Thread.sleep(time)
        sendAll(msg)
    }

    private fun parseValues() {
        val raw_values = Config.mob_filter!!.get()
        for (s in raw_values) {
            val mob_name = s.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
            val worlds = s.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1].split(",".toRegex())
                .dropLastWhile { it.isEmpty() }
                .toTypedArray()
            mobFilter[mob_name] = Arrays.asList(*worlds)
        }
    }

    fun adventure(): MinecraftServerAudiences {
        checkNotNull(this.adventure) { "Tried to access Adventure without a running server!" }
        return this.adventure!!
    }
}